File: tests/algorithms/x/PHP/data_structures/queues/double_ended_queue.php
<?php
ini_set('memory_limit', '-1');
function _len($x) {
    if (is_array($x)) { return count($x); }
    if (is_string($x)) { return strlen($x); }
    return strlen(strval($x));
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
function empty_deque() {
  return ['data' => []];
}
function push_back($dq, $value) {
  return ['data' => _append($dq['data'], $value)];
}
function push_front($dq, $value) {
  $res = [$value];
  $i = 0;
  while ($i < _len($dq['data'])) {
  $res = _append($res, $dq['data'][$i]);
  $i = $i + 1;
};
  return ['data' => $res];
}
function extend_back($dq, $values) {
  $res = $dq['data'];
  $i = 0;
  while ($i < count($values)) {
  $res = _append($res, $values[$i]);
  $i = $i + 1;
};
  return ['data' => $res];
}
function extend_front($dq, $values) {
  $res = [];
  $i = count($values) - 1;
  while ($i >= 0) {
  $res = _append($res, $values[$i]);
  $i = $i - 1;
};
  $j = 0;
  while ($j < _len($dq['data'])) {
  $res = _append($res, $dq['data'][$j]);
  $j = $j + 1;
};
  return ['data' => $res];
}
function pop_back($dq) {
  if (_len($dq['data']) == 0) {
  $panic('pop from empty deque');
}
  $res = [];
  $i = 0;
  while ($i < _len($dq['data']) - 1) {
  $res = _append($res, $dq['data'][$i]);
  $i = $i + 1;
};
  return ['deque' => ['data' => $res], 'value' => $dq['data'][_len($dq['data']) - 1]];
}
function pop_front($dq) {
  if (_len($dq['data']) == 0) {
  $panic('popleft from empty deque');
}
  $res = [];
  $i = 1;
  while ($i < _len($dq['data'])) {
  $res = _append($res, $dq['data'][$i]);
  $i = $i + 1;
};
  return ['deque' => ['data' => $res], 'value' => $dq['data'][0]];
}
function is_empty($dq) {
  return _len($dq['data']) == 0;
}
function length($dq) {
  return _len($dq['data']);
}
function to_string($dq) {
  if (_len($dq['data']) == 0) {
  return '[]';
}
  $s = '[' . _str($dq['data'][0]);
  $i = 1;
  while ($i < _len($dq['data'])) {
  $s = $s . ', ' . _str($dq['data'][$i]);
  $i = $i + 1;
};
  return $s . ']';
}
function main() {
  $dq = empty_deque();
  $dq = push_back($dq, 2);
  $dq = push_front($dq, 1);
  $dq = extend_back($dq, [3, 4]);
  $dq = extend_front($dq, [0]);
  echo rtrim(to_string($dq)), PHP_EOL;
  $r = pop_back($dq);
  $dq = $r['deque'];
  echo rtrim(json_encode($r['value'], 1344)), PHP_EOL;
  $r = pop_front($dq);
  $dq = $r['deque'];
  echo rtrim(json_encode($r['value'], 1344)), PHP_EOL;
  echo rtrim(to_string($dq)), PHP_EOL;
  echo rtrim(json_encode(is_empty(empty_deque()), 1344)), PHP_EOL;
}
main();

File: tests/algorithms/transpiler/FS/data_structures/queues/double_ended_queue.fs
// Generated 2025-08-07 14:57 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
type Deque = {
    data: int array
}
type PopResult = {
    deque: Deque
    value: int
}
let rec empty_deque () =
    let mutable __ret : Deque = Unchecked.defaultof<Deque>
    try
        __ret <- { data = [||] }
        raise Return
        __ret
    with
        | Return -> __ret
and push_back (dq: Deque) (value: int) =
    let mutable __ret : Deque = Unchecked.defaultof<Deque>
    let mutable dq = dq
    let mutable value = value
    try
        __ret <- { data = Array.append (dq.data) [|value|] }
        raise Return
        __ret
    with
        | Return -> __ret
and push_front (dq: Deque) (value: int) =
    let mutable __ret : Deque = Unchecked.defaultof<Deque>
    let mutable dq = dq
    let mutable value = value
    try
        let mutable res: int array = [|value|]
        let mutable i: int = 0
        while i < (Seq.length (dq.data)) do
            res <- Array.append res [|_idx (dq.data) (i)|]
            i <- i + 1
        __ret <- { data = res }
        raise Return
        __ret
    with
        | Return -> __ret
and extend_back (dq: Deque) (values: int array) =
    let mutable __ret : Deque = Unchecked.defaultof<Deque>
    let mutable dq = dq
    let mutable values = values
    try
        let mutable res: int array = dq.data
        let mutable i: int = 0
        while i < (Seq.length (values)) do
            res <- Array.append res [|_idx values (i)|]
            i <- i + 1
        __ret <- { data = res }
        raise Return
        __ret
    with
        | Return -> __ret
and extend_front (dq: Deque) (values: int array) =
    let mutable __ret : Deque = Unchecked.defaultof<Deque>
    let mutable dq = dq
    let mutable values = values
    try
        let mutable res: int array = [||]
        let mutable i: int = (Seq.length (values)) - 1
        while i >= 0 do
            res <- Array.append res [|_idx values (i)|]
            i <- i - 1
        let mutable j: int = 0
        while j < (Seq.length (dq.data)) do
            res <- Array.append res [|_idx (dq.data) (j)|]
            j <- j + 1
        __ret <- { data = res }
        raise Return
        __ret
    with
        | Return -> __ret
and pop_back (dq: Deque) =
    let mutable __ret : PopResult = Unchecked.defaultof<PopResult>
    let mutable dq = dq
    try
        if (Seq.length (dq.data)) = 0 then
            failwith ("pop from empty deque")
        let mutable res: int array = [||]
        let mutable i: int = 0
        while i < ((Seq.length (dq.data)) - 1) do
            res <- Array.append res [|_idx (dq.data) (i)|]
            i <- i + 1
        __ret <- { deque = { data = res }; value = _idx (dq.data) ((Seq.length (dq.data)) - 1) }
        raise Return
        __ret
    with
        | Return -> __ret
and pop_front (dq: Deque) =
    let mutable __ret : PopResult = Unchecked.defaultof<PopResult>
    let mutable dq = dq
    try
        if (Seq.length (dq.data)) = 0 then
            failwith ("popleft from empty deque")
        let mutable res: int array = [||]
        let mutable i: int = 1
        while i < (Seq.length (dq.data)) do
            res <- Array.append res [|_idx (dq.data) (i)|]
            i <- i + 1
        __ret <- { deque = { data = res }; value = _idx (dq.data) (0) }
        raise Return
        __ret
    with
        | Return -> __ret
and is_empty (dq: Deque) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable dq = dq
    try
        __ret <- (Seq.length (dq.data)) = 0
        raise Return
        __ret
    with
        | Return -> __ret
and length (dq: Deque) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable dq = dq
    try
        __ret <- Seq.length (dq.data)
        raise Return
        __ret
    with
        | Return -> __ret
and to_string (dq: Deque) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable dq = dq
    try
        if (Seq.length (dq.data)) = 0 then
            __ret <- "[]"
            raise Return
        let mutable s: string = "[" + (_str (_idx (dq.data) (0)))
        let mutable i: int = 1
        while i < (Seq.length (dq.data)) do
            s <- (s + ", ") + (_str (_idx (dq.data) (i)))
            i <- i + 1
        __ret <- s + "]"
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let mutable dq: Deque = empty_deque()
        dq <- push_back (dq) (2)
        dq <- push_front (dq) (1)
        dq <- extend_back (dq) (unbox<int array> [|3; 4|])
        dq <- extend_front (dq) (unbox<int array> [|0|])
        printfn "%s" (to_string (dq))
        let mutable r: PopResult = pop_back (dq)
        dq <- r.deque
        printfn "%d" (r.value)
        r <- pop_front (dq)
        dq <- r.deque
        printfn "%d" (r.value)
        printfn "%s" (to_string (dq))
        printfn "%b" (is_empty (empty_deque()))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
